ctenex/domain/models/matching_engine.py

```diff
@@ -0,0 +1,179 @@
+
+        if len(trades) > 0:
+            logger.debug(f"Matched order with ID {buy_order.id}")
+            logger.debug(f"Generated {len(trades)} trades:")
+            for trade in trades:
+                logger.debug(trade)
+
+        return trades
+
+    def _match_sell_order(self, sell_order: Order) -> list[Trade]:
+        trades = []
+        order_book = self.order_books[sell_order.contract_id]
+
+        while sell_order.remaining_quantity and sell_order.remaining_quantity > 0:
+            # Check if there are any bids to match against
+            if not order_book.bids or not order_book.bid_queues:
+                break
+
+            best_bid_price = -order_book.bids.keys()[0]  # Convert back from negative
+
+            # For limit orders, check if the price is acceptable
+            if (
+                sell_order.order_type == OrderType.LIMIT
+                and best_bid_price < sell_order.price
+            ):
+                break
+
+            # Match against the best bid price
+            bid_queue = order_book.bid_queues[best_bid_price]
+            while bid_queue and sell_order.remaining_quantity > 0:
+                buy_order = bid_queue[0]
+
+                # Calculate trade quantity
+                trade_quantity = min(
+                    sell_order.remaining_quantity, buy_order.remaining_quantity
+                )
+
+                # Create and record the trade
+                trade = Trade(
+                    contract_id=sell_order.contract_id,
+                    buy_order_id=buy_order.id,
+                    sell_order_id=sell_order.id,
+                    price=best_bid_price,
+                    quantity=trade_quantity,
+                )
+                trades.append(trade)
+
+                # Update order quantities
+                sell_order.remaining_quantity -= trade_quantity
+                buy_order.remaining_quantity -= trade_quantity
+
+                # Update order statuses
+                if buy_order.remaining_quantity == 0:
+                    buy_order.status = OrderStatus.FILLED
+                    bid_queue.pop(0)
+                    order_book.orders_by_id.pop(buy_order.id)
+                else:
+                    buy_order.status = OrderStatus.PARTIALLY_FILLED
+
+                if sell_order.remaining_quantity == 0:
+                    sell_order.status = OrderStatus.FILLED
+                else:
+                    sell_order.status = OrderStatus.PARTIALLY_FILLED
+
+            # If bid queue is empty, remove the price level
+            if not bid_queue:
+                order_book.bid_queues.pop(best_bid_price)
+                order_book.bids.pop(-best_bid_price)
+
+        # Temp
+        if len(trades) > 0:
+            logger.debug(f"Matched order with ID {sell_order.id}")
+            logger.debug(f"Generated {len(trades)} trades:")
+            for trade in trades:
+                logger.debug(trade)
+
+        return trades
```

Comment:
Cool. Some order book visualisation method is useful for traders and analysts. Could be something basic like: 

```
def get_order_book_snapshot(self, contract_id: ContractCode, depth: int = 5) -> dict:
    """Return a structured view of the current order book to a specified depth."""
    book = self.order_books[contract_id]
    return {
        "bids": [{"price": p, "quantity": sum(o.remaining_quantity for o in q)} 
                for p, q in sorted([(float(-k), v) for k, v in book.bids.items()], reverse=True)[:depth]],
        "asks": [{"price": p, "quantity": sum(o.remaining_quantity for o in q)} 
                for p, q in sorted([(float(k), v) for k, v in book.asks.items()])[:depth]]
    }
```